options/options_helper.cc

```diff
@@ -1015,28 +916,26 @@ Status StringToMap(const std::string& opts_str,
 }
 
 Status GetStringFromStruct(
-    const ConfigOptions& cfg_options, const void* const opt_ptr,
+    const ConfigOptions& config_options, const void* const opt_ptr,
     const std::unordered_map<std::string, OptionTypeInfo>& type_info,
     std::string* opt_string) {
   assert(opt_string);
   opt_string->clear();
   for (const auto iter : type_info) {
     const auto& opt_info = iter.second;
-    if (opt_info.IsDeprecated()) {
-      // If the option is no longer used in rocksdb and marked as deprecated,
-      // we skip it in the serialization.
-      continue;
-    }
-    const char* opt_address =
-        reinterpret_cast<const char*>(opt_ptr) + opt_info.offset;
-    std::string value;
-    bool result =
-        SerializeSingleOptionHelper(opt_address, opt_info.type, &value);
-    if (result) {
-      opt_string->append(iter.first + "=" + value + cfg_options.delimiter);
-    } else {
-      return Status::InvalidArgument("failed to serialize %s\n",
-                                     iter.first.c_str());
+    // If the option is no longer used in rocksdb and marked as deprecated,
+    // we skip it in the serialization.
+    if (opt_info.ShouldSerialize()) {
```

Comment:
Add comments to explicitly explain the case of IsDeprecated(). Also, if !IsDeprecated() but IsAlias() or IsEnabled(OptionTypeFlags::kStringNone), ShouldSerialize() also returns False, has the behavior been changed?

Reply:
IsAlias and IsEnabled(kStringNone) are new.  There were some options previously that were listed in the "parse" code but not in the "serialize" code.  When I moved them into the OptionTypeInfo, I marked them as "kStringNone".

IsAlias is also new and is used when there are multiple names referring to the same potential object -- like "memtable" and "memtable_factory" or "table_factory" and "block_based_table_factory".  In those cases, we do not want to serialize the alias as it will be handled by the definition.  

I will add a comment in "ShouldSerialize" explaining this intended behavior

Reply:
> IsAlias and IsEnabled(kStringNone) are new. There were some options previously that were listed in the "parse" code but not in the "serialize" code. When I moved them into the OptionTypeInfo, I marked them as "kStringNone".
> 
> IsAlias is also new and is used when there are multiple names referring to the same potential object -- like "memtable" and "memtable_factory" or "table_factory" and "block_based_table_factory". In those cases, we do not want to serialize the alias as it will be handled by the definition.
> 
> I will add a comment in "ShouldSerialize" explaining this intended behavior

Maybe use another name instead of "kStringNone", it is hard to understand.

Reply:
Renamed to kDontSerialize